src/main/java/javaslang/collection/HashArrayMappedTrie.java

```diff
@@ -376,26 +355,49 @@ V value() {
                 if (action == REMOVE) {
                     return filtered;
                 } else {
-                    if (filtered.isEmpty()) {
-                        return new LeafSingleton<>(hash, key, value);
```

Comment:
I'm not as deep involved as you - but this catched my eye - do we need this singleton any more here?